server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -457,20 +458,24 @@ public Builder removeReplica() {
                 for (ShardRouting shardRouting : indexShard) {
                     builder.addShard(shardRouting);
                 }
-                // first check if there is one that is not assigned to a node, and remove it
+
+                // define remove priority
+                List<Function<ShardRouting, Boolean>> priorityRemoveClauses = List.of(
+                    shardRouting -> shardRouting.assignedToNode() == false,
+                    shardRouting -> shardRouting.initializing(),
+                    shardRouting -> shardRouting.relocating(),
```

Comment:
I don't see a good reason to prefer dropping relocating shards over stationary ones. The shard might be relocating because e.g. an allocation filter was applied, and nearly finished, and in that case we would drop it and lose all that progress and then immediately start relocating another copy.

Reply:
This is indeed a case worth considering! But I can also think of some cases in which removing relocating ones is better. For example, a node reaches watermark and elasticsearch triggered the relocating automatically, then the user expects to release the disk space by removing some replicas. In this case we should remove the relocating ones to release the disk space. Actually, apart from shard status, there can be many factors affecting the removing priority: allocation filter, node disk space, shard balance... maybe we can apply all of them well in the feature but that's a big job.
In this PR, we just consider in the term of shard status. We can see removing relocating shards in a higher priority can be sometimes good sometimes bad. But i think removing replicas in an consistent logic (instead of randomly) is important because at least user can know which replica will be removed before he post a request. In the case you mentioned, the user can post the `put-setting` request after the relocating one completed and the another one started relocating, and he can make sure the relocating one will be removed now. What do you think?

Reply:
We are still removing replicas randomly even with these changes, because for instance we don't specify _which_ relocating copy to remove, there may be multiple, and only one of them might be the right one. It certainly won't be easy to address that, nor do we need to address it here. However, it's a blocker on this PR to prefer relocating shards as you have proposed since it sometimes makes things worse.

Reply:
@DaveCTurner Thank you very much for the explanation! I have fixed this.